File: backend/domain/src/main/java/com/system/alba/model/dto/TokenDto.java
package com.system.alba.model.dto;

import com.system.alba.common.AppType;
import com.system.alba.common.GenericMapper;
import com.system.alba.model.domain.Token;
import jakarta.validation.constraints.NotBlank;
import lombok.Getter;
import lombok.Setter;
import org.mapstruct.ReportingPolicy;
import org.mapstruct.factory.Mappers;

import java.time.LocalDateTime;
import java.util.List;

public class TokenDto {

    @Getter
    @Setter
    public static class Detail extends Summary {
        private AccountDto.Abbr account;

        @org.mapstruct.Mapper(unmappedTargetPolicy = ReportingPolicy.IGNORE)
        public interface Mapper extends GenericMapper<Detail, Token> {
            Mapper INSTANCE = Mappers.getMapper(Mapper.class);
        }
    }

    @Getter
    @Setter
    public static class Summary {
        private String accessToken;
        private String refreshToken;
        private List<AppType.AccountRole> roles;
        private LocalDateTime accessExpiresAt;
        private LocalDateTime refreshExpiresAt;
    }

    @Getter
    @Setter
    public static class TokenResponse {
        private String accessToken;
        private int accessTokenExpiresIn;
        private List<AppType.AccountRole> roles;
        private AccountDto.Abbr account;

        @org.mapstruct.Mapper(unmappedTargetPolicy = ReportingPolicy.IGNORE)
        public interface Mapper extends GenericMapper<TokenResponse, Detail> {
            Mapper INSTANCE = Mappers.getMapper(Mapper.class);
        }
    }

    @Getter
    @Setter
    public static class AppTokenResponse {
        private String accessToken;
        private int accessTokenExpiresIn;
        private String refreshToken;
        private int refreshTokenExpiresIn;
        private List<AppType.AccountRole> roles;
        private AccountDto.Abbr account;

        @org.mapstruct.Mapper(unmappedTargetPolicy = ReportingPolicy.IGNORE)
        public interface Mapper extends GenericMapper<AppTokenResponse, Detail> {
            Mapper INSTANCE = Mappers.getMapper(Mapper.class);
        }
    }

    @Getter
    @Setter
    public static class RefreshRequest {
        @NotBlank
        private String refreshToken;
    }

    @Getter
    @Setter
    public static class LogoutRequest {
        @NotBlank
        private String refreshToken;
    }

}

File: mobile-monorepo/packages/common/lib/src/model/dto/token_dto.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'token_dto.freezed.dart';

@freezed
class TokenResponse with _$TokenResponse {
  const factory TokenResponse({
    required String accessToken,
    required int accessTokenExpiresIn,
    required String refreshToken,
    required int refreshTokenExpiresIn,
    required String userId,
    required String email,
    required List<String> roles,
  }) = _TokenResponse;

  factory TokenResponse.fromJson(Map<String, dynamic> json) {
    final data = json['data'] is Map<String, dynamic>
        ? json['data'] as Map<String, dynamic>
        : json;
    final account = data['account'] is Map<String, dynamic>
        ? data['account'] as Map<String, dynamic>
        : const <String, dynamic>{};

    return TokenResponse(
      accessToken: data['accessToken'] as String? ?? '',
      accessTokenExpiresIn: data['accessTokenExpiresIn'] as int? ?? 0,
      refreshToken: data['refreshToken'] as String? ?? '',
      refreshTokenExpiresIn: data['refreshTokenExpiresIn'] as int? ?? 0,
      userId: account['id'] as String? ?? '',
      email: account['email'] as String? ?? '',
      roles: (data['roles'] as List<dynamic>? ?? const <dynamic>[])
          .whereType<String>()
          .toList(),
    );
  }
}
